File: leetcode/297.SerializeAndDeserializeBinaryTree.go
package leetcode

import (
	"strconv"
	"strings"
)

type Codec struct {
}

// Serializes a tree to a single string.
func (c *Codec) serialize(root *TreeNode) string {
	var traverse func(node *TreeNode)

	var result []string

	traverse = func(node *TreeNode) {
		if node == nil {
			result = append(result, "nil")
			return
		}

		result = append(result, strconv.Itoa(node.Val))
		traverse(node.Left)
		traverse(node.Right)
	}

	traverse(root)

	return strings.Join(result, ",")
}

// Deserializes your encoded data to tree.
func (c *Codec) deserialize(data string) *TreeNode {
	var traverse func() *TreeNode
	values := strings.Split(data, ",")

	i := 0
	traverse = func() *TreeNode {
		if string(values[i]) == "nil" {
			i++
			return nil
		}

		val, _ := strconv.Atoi(values[i])
		node := &TreeNode{Val: val}
		i++
		node.Left = traverse()
		node.Right = traverse()

		return node
	}

	return traverse()
}

File: leetcode/297.SerializeAndDeserializeBinaryTree.ts
import { TreeNode } from './shared'

function serialize(root: TreeNode | null): string {
  const result: string[] = []

  const dfs = (node: TreeNode | null) => {
    if (!node) {
      result.push('null')
      return
    }

    result.push(String(node.val))
    dfs(node.left)
    dfs(node.right)
  }

  dfs(root)

  return result.join(',')
}

/*
 * Decodes your encoded data to tree.
 */
function deserialize(data: string): TreeNode | null {
  const values = data.split(',')
  let i = 0

  const dfs = (): TreeNode | null => {
    if (values[i] === 'null') {
      i++
      return null
    }

    const node = new TreeNode(Number(values[i]))
    i++
    node.left = dfs()
    node.right = dfs()

    return node
  }

  return dfs()
}

const tree = new TreeNode(1)
tree.left = new TreeNode(2)
tree.right = new TreeNode(3)
tree.right.left = new TreeNode(4)
tree.right.right = new TreeNode(5)

const serialized = serialize(tree)
console.log(serialized)
console.log(deserialize(serialized))
